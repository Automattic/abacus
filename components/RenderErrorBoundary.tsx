import React from 'react'

interface RenderError {
  clear: () => void
  error: Error
  info: React.ErrorInfo
}

type ChildrenParam0 = {
  renderError: RenderError | null
}

type Props = {
  children: (arg0: ChildrenParam0) => React.ReactNode
  onClear?: () => void
  onError?: (renderError: RenderError) => void
}

type State = {
  error: Error | null
  info: React.ErrorInfo | null
}

/**
 * A render error boundary component.
 *
 * Renders its render prop with an optional `renderError` prop. If `renderError`
 * is not `null`, then the render prop should handle appropriately for the best UX.
 *
 *     import RenderErrorBoundary from '.../RenderErrorBoundary'
 *
 *     const MyComponent = () => {
 *       const handleRenderError = (renderError) => {
 *         const { error, info } = renderError
 *         console.error('Render Error:\n\n', error, info)
 *       }
 *
 *       return (
 *         <RenderErrorBoundary onError={handleRenderError}>
 *           {({ renderError }) => (
 *             <React.Fragment>
 *               {renderError
 *                 ? <RenderErrorNotification renderError={renderError} />
 *                 : <Etc />
 *               }
 *             </React.Fragment>
 *           )}
 *         </RenderErrorBoundary>
 *       )
 *     }
 *
 * A `renderError` has a `clear` function that can be used to clear the error.
 *
 *     const RenderErrorNotification = ({ renderError }) => {
 *       return (
 *         <div>
 *           <h1>An error occurred!</h1>
 *           <p>{renderError.error.message}</p>
 *           <button onClick={renderError.clear}>Clear</button>
 *         </div>
 *       )
 *     }
 *
 * @see https://reactjs.org/docs/error-boundaries.htm
 */
class RenderErrorBoundary extends React.PureComponent<Props, State> {
  state = { error: null, info: null }

  clear = () => {
    this.setState(
      {
        error: null,
        info: null,
      },
      () => {
        if (this.props.onClear) {
          this.props.onClear()
        }
      },
    )
  }

  componentDidCatch(error: Error, info: React.ErrorInfo) {
    this.setState({ error, info })
    if (this.props.onError) {
      const renderError = {
        clear: this.clear,
        error,
        info,
      }
      this.props.onError(renderError)
    }
  }

  render() {
    const { error, info } = this.state
    const childrenArg = {
      renderError: (error && info
        ? {
            clear: this.clear,
            error,
            info,
          }
        : null) as RenderError | null,
    }
    return this.props.children(childrenArg)
  }
}

export type { RenderError }

export default RenderErrorBoundary
