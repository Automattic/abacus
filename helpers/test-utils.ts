import mediaQuery from 'css-mediaquery'

/**
 * Create a `matchMedia` function that will match a query based on the specified
 * width.
 *
 *     const initialJsDomWindowInnerWidth = window.innerWidth
 *     afterEach(() => {
 *       // Reset back to initial width for tests that don't explicitly set the width
 *       // themselves.
 *       window.matchMedia = createMatchMedia(initialJsDomWindowInnerWidth)
 *     })
 *
 *     test('...', () => {
 *       window.matchMedia = createMatchMedia(600)
 *       ...
 *     })
 *
 * @param width - The width of the window to simulate.
 */
function createMatchMedia(width: number) {
  return (query: string) => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    addEventListener: jest.fn(),
    addListener: jest.fn(),
    dispatchEvent: jest.fn(),
    onchange: jest.fn(),
    removeEventListener: jest.fn(),
    removeListener: jest.fn(),
  })
}

export { createMatchMedia }
