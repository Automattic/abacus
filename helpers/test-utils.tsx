import { Queries, render as actualRender, RenderOptions } from '@testing-library/react'
import mediaQuery from 'css-mediaquery'
import React from 'react'

import ThemeProvider from '@/styles/ThemeProvider'

/**
 * A wrapped unit-test react-renderer, useful for adding React Contexts globally.
 */
export const render: typeof actualRender = <Q extends Queries>(ui: React.ReactElement, options?: RenderOptions<Q>) =>
  actualRender((<ThemeProvider>{ui}</ThemeProvider>) as React.ReactElement, options) as ReturnType<typeof actualRender>

/**
 * Create a `matchMedia` function that will match a query based on the specified
 * width.
 *
 *     const initialJsDomWindowInnerWidth = window.innerWidth
 *     afterEach(() => {
 *       // Reset back to initial width for tests that don't explicitly set the width
 *       // themselves.
 *       window.matchMedia = createMatchMedia(initialJsDomWindowInnerWidth)
 *     })
 *
 *     test('...', () => {
 *       window.matchMedia = createMatchMedia(600)
 *       ...
 *     })
 *
 * @param width - The width of the window to simulate.
 */
export function createMatchMedia(width: number) {
  return (query: string) => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    addEventListener: jest.fn(),
    addListener: jest.fn(),
    dispatchEvent: jest.fn(),
    onchange: jest.fn(),
    removeEventListener: jest.fn(),
    removeListener: jest.fn(),
  })
}
