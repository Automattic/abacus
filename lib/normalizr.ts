import { normalize } from 'normalizr'

import { MetricBare, metricBareNormalizrSchema, MetricFull, Segment, segmentNormalizrSchema } from './schemas'

export function indexMetrics<Metric extends MetricBare | MetricFull>(metrics: Metric[]) {
  const {
    entities: { metrics: indexedMetrics },
  } = normalize<Metric>(metrics, [metricBareNormalizrSchema])
  /* istanbul ignore next */
  if (!indexedMetrics) {
    throw new Error(`No metrics produced after normalisation, this should never happen.`)
  }
  return indexedMetrics
}

export function indexSegments(segments: Segment[]) {
  const {
    entities: { segments: indexedSegments },
  } = normalize<Segment>(segments, [segmentNormalizrSchema])
  /* istanbul ignore next */
  if (!indexedSegments) {
    throw new Error(`No segments produced after normalisation, this should never happen.`)
  }
  return indexedSegments
}
