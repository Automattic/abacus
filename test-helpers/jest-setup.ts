import { format } from 'util'

// Throw errors on any console logging.
// Adapted from https://github.com/facebook/jest/issues/6121#issuecomment-529591574
;['error', 'info', 'log', 'warn'].forEach((funcName) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(global.console as any)[funcName] = function (message?: any, ...optionalParams: any[]) {
    throw new Error(format(message, optionalParams))
  }
})
