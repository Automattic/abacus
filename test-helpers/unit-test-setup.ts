/* eslint-disable @typescript-eslint/ban-ts-ignore */
// @ts-ignore
window.URL.createObjectURL = jest.fn()

// @ts-ignore
window.HTMLCanvasElement.prototype.getContext = jest.fn()
