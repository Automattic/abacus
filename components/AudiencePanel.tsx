import { createStyles, makeStyles, Paper, Toolbar, Typography } from '@material-ui/core'
import { TableCellProps } from '@material-ui/core/TableCell'
import _ from 'lodash'
import React, { useMemo } from 'react'

import LabelValueTable from '@/components/LabelValueTable'
import SegmentsTable from '@/components/SegmentsTable'
import VariationsTable from '@/components/VariationsTable'
import { ExperimentFull, Segment, SegmentAssignment, SegmentType } from '@/lib/schemas'
import * as Variations from '@/lib/variations'
import theme from '@/styles/theme'

/**
 * Resolves the segment ID of the segment assignment with the actual segment.
 * If the ID cannot be resolved, then an `Error` will be thrown.
 *
 * @param segmentAssignments - The segment assignments to be resolved.
 * @param segments - The segments to associate with the assignments.
 * @throws {Error} When unable to resolve a segment ID with one of the supplied
 *   segments.
 */
function resolveSegmentAssignments(
  segmentAssignments: SegmentAssignment[],
  segments: Segment[],
): {
  segment: Segment
  isExcluded: boolean
}[] {
  const segmentsById: { [segmentId: string]: Segment } = {}
  segments.forEach((segment) => (segmentsById[segment.segmentId] = segment))

  return segmentAssignments.map((segmentAssignment) => {
    const segment = segmentsById[segmentAssignment.segmentId]

    if (!segment) {
      throw Error(
        `Failed to lookup segment with ID ${segmentAssignment.segmentId} for assignment with ID ${segmentAssignment.segmentAssignmentId}.`,
      )
    }

    return {
      segment,
      isExcluded: segmentAssignment.isExcluded,
    }
  })
}

const useStyles = makeStyles(() =>
  createStyles({
    title: {
      flexGrow: 1,
    },
  }),
)

const eventStyles = makeStyles(() =>
  createStyles({
    entry: {
      display: 'block',
      fontFamily: theme.custom.fonts.monospace,
      color: 'gray',
    },
    eventName: {
      fontFamily: theme.custom.fonts.monospace,
    },
    propsList: {
      margin: 0,
      paddingInlineStart: 0,
    },
    eventList: {
      '& p:not(:first-child)': {
        paddingTop: theme.spacing(2),
      },
      '& p': {
        paddingBottom: theme.spacing(2),
      },
      '& p:not(:last-child)': {
        borderBottom: '1px solid rgb(224,224,224)',
      },
    },
  }),
)

function ExposureEventsTable({ experiment }: { experiment: ExperimentFull }) {
  const classes = eventStyles()

  return (
    <div className={classes.eventList}>
      {experiment.exposureEvents?.map((ev) => (
        <Typography key={ev.event}>
          <span className={classes.eventName}>{ev.event}</span>
          {ev.props && (
            <ul className={classes.propsList}>
              {' '}
              {Object.entries(ev.props).map(([key, val]) => (
                <li key={key + val} className={classes.entry}>
                  {key}: {val}
                </li>
              ))}
            </ul>
          )}
        </Typography>
      )) ?? <Typography>No exposure events defined</Typography>}
    </div>
  )
}

/**
 * Renders the audience information of an experiment in a panel component.
 *
 * @param props.experiment - The experiment with the audience information.
 * @param props.segments - The segments to look up (aka resolve) the segment IDs
 *   of the experiment's segment assignments.
 */
function AudiencePanel({ experiment, segments }: { experiment: ExperimentFull; segments: Segment[] }) {
  const classes = useStyles()

  const segmentsByType = useMemo(
    () => _.groupBy(resolveSegmentAssignments(experiment.segmentAssignments, segments), _.property('segment.type')),
    [experiment.segmentAssignments, segments],
  )

  const data = [
    { label: 'Platform', value: experiment.platform },
    { label: 'User Type', value: experiment.existingUsersAllowed ? 'All users (new + existing)' : 'New users only' },
    {
      label: 'Variations',
      padding: 'none' as TableCellProps['padding'],
      value: <VariationsTable variations={Variations.sort(experiment.variations)} />,
    },
    {
      label: 'Segments',
      padding: 'none' as TableCellProps['padding'],
      value: (
        <>
          <SegmentsTable
            resolvedSegmentAssignments={segmentsByType[SegmentType.Locale] ?? []}
            type={SegmentType.Locale}
          />
          <SegmentsTable
            resolvedSegmentAssignments={segmentsByType[SegmentType.Country] ?? []}
            type={SegmentType.Country}
          />
        </>
      ),
    },
    {
      label: 'Exposure Events',
      value: <ExposureEventsTable experiment={experiment} />,
    },
  ]
  return (
    <Paper>
      <Toolbar>
        <Typography className={classes.title} color='textPrimary' variant='h3'>
          Audience
        </Typography>
      </Toolbar>
      <LabelValueTable data={data} />
    </Paper>
  )
}

export default AudiencePanel
