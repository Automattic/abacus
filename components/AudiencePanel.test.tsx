import { normalize } from 'normalizr'
import React from 'react'

import RenderErrorBoundary from '@/components/RenderErrorBoundary'
import {
  ExperimentFull,
  ExperimentFullNormalizedEntities,
  experimentFullNormalizrSchema,
  Segment,
  segmentNormalizrSchema,
} from '@/lib/schemas'
import Fixtures from '@/test-helpers/fixtures'
import { render } from '@/test-helpers/test-utils'

import AudiencePanel from './AudiencePanel'

test('renders as expected with no segment assignments', () => {
  const experiment = Fixtures.createExperimentFull({
    segmentAssignments: [],
  })
  const normalizedExperimentData = normalize<ExperimentFull, ExperimentFullNormalizedEntities>(
    experiment,
    experimentFullNormalizrSchema,
  )
  const { container } = render(
    <AudiencePanel normalizedExperimentData={normalizedExperimentData} indexedSegments={{}} />,
  )

  expect(container).toMatchSnapshot()
})

test('renders as expected with existing users allowed', () => {
  const experiment = Fixtures.createExperimentFull({
    existingUsersAllowed: true,
    segmentAssignments: [],
  })
  const normalizedExperimentData = normalize<ExperimentFull, ExperimentFullNormalizedEntities>(
    experiment,
    experimentFullNormalizrSchema,
  )
  const { container } = render(
    <AudiencePanel normalizedExperimentData={normalizedExperimentData} indexedSegments={{}} />,
  )

  expect(container).toMatchSnapshot()
})

test('renders as expected with all segments resolvable', () => {
  const segments = Fixtures.createSegments(5)
  const experiment = Fixtures.createExperimentFull({
    segmentAssignments: [
      Fixtures.createSegmentAssignment({ segmentAssignmentId: 101, segmentId: 1 }),
      Fixtures.createSegmentAssignment({ segmentAssignmentId: 102, segmentId: 2, isExcluded: true }),
      Fixtures.createSegmentAssignment({ segmentAssignmentId: 103, segmentId: 3, isExcluded: true }),
      Fixtures.createSegmentAssignment({ segmentAssignmentId: 104, segmentId: 4 }),
    ],
  })
  const normalizedExperimentData = normalize<ExperimentFull, ExperimentFullNormalizedEntities>(
    experiment,
    experimentFullNormalizrSchema,
  )
  const {
    entities: { segments: indexedSegments },
  } = normalize<Segment, { segments: Record<number, Segment> }>(segments, [segmentNormalizrSchema])
  const { container } = render(
    <AudiencePanel normalizedExperimentData={normalizedExperimentData} indexedSegments={indexedSegments} />,
  )

  expect(container).toMatchSnapshot()
})

test('throws an error when some segments not resolvable', () => {
  const segments = Fixtures.createSegments(5)
  const experiment = Fixtures.createExperimentFull({
    segmentAssignments: [
      Fixtures.createSegmentAssignment({ segmentAssignmentId: 101, segmentId: 1 }),
      Fixtures.createSegmentAssignment({ segmentAssignmentId: 102, segmentId: 2, isExcluded: true }),
      Fixtures.createSegmentAssignment({ segmentAssignmentId: 103, segmentId: 3, isExcluded: true }),
      Fixtures.createSegmentAssignment({ segmentAssignmentId: 104, segmentId: 4 }),
      Fixtures.createSegmentAssignment({ segmentAssignmentId: 110, segmentId: 10 }),
    ],
  })

  const normalizedExperimentData = normalize<ExperimentFull, ExperimentFullNormalizedEntities>(
    experiment,
    experimentFullNormalizrSchema,
  )
  const {
    entities: { segments: indexedSegments },
  } = normalize<Segment, { segments: Record<number, Segment> }>(segments, [segmentNormalizrSchema])

  // Note: This console.error spy is mainly used to suppress the output that the
  // `render` function outputs.
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const consoleErrorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => {})
  try {
    render(
      <RenderErrorBoundary>
        {() => <AudiencePanel normalizedExperimentData={normalizedExperimentData} indexedSegments={indexedSegments} />}
      </RenderErrorBoundary>,
    )
    expect(false).toBe(true) // Should never be reached
  } catch (err) {
    expect(consoleErrorSpy).toHaveBeenCalled()
  } finally {
    consoleErrorSpy.mockRestore()
  }
})
