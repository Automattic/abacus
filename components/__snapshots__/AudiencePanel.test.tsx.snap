// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected with all segments resolvable 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <h3
      class="MuiTypography-root makeStyles-title-5 MuiTypography-h3 MuiTypography-colorTextPrimary"
    >
      Audience
    </h3>
    <table
      class="MuiTable-root"
    >
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Platform
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            calypso
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            User Type
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            New users only
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Variations
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
          >
            <table
              class="MuiTable-root"
            >
              <thead
                class="MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    role="columnheader"
                    scope="col"
                  >
                    Name
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    role="columnheader"
                    scope="col"
                  >
                    Percent
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root"
              >
                <tr
                  class="MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    test
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    40
                    %
                  </td>
                </tr>
                <tr
                  class="MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    control
                    <div
                      class="MuiChip-root makeStyles-default-6"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        Default
                      </span>
                    </div>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    60
                    %
                  </td>
                </tr>
              </tbody>
            </table>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Segments
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
          >
            <table
              class="MuiTable-root"
            >
              <thead
                class="MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    role="columnheader"
                    scope="col"
                  >
                    Countries
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root"
              >
                <tr
                  class="MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    segment_2
                    <div
                      class="MuiChip-root makeStyles-excluded-7 MuiChip-sizeSmall"
                    >
                      <span
                        class="MuiChip-label MuiChip-labelSmall"
                      >
                        Excluded
                      </span>
                    </div>
                  </td>
                </tr>
                <tr
                  class="MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    segment_4
                  </td>
                </tr>
              </tbody>
            </table>
            <table
              class="MuiTable-root"
            >
              <thead
                class="MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    role="columnheader"
                    scope="col"
                  >
                    Locales
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root"
              >
                <tr
                  class="MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    segment_1
                  </td>
                </tr>
                <tr
                  class="MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    segment_3
                    <div
                      class="MuiChip-root makeStyles-excluded-7 MuiChip-sizeSmall"
                    >
                      <span
                        class="MuiChip-label MuiChip-labelSmall"
                      >
                        Excluded
                      </span>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`renders as expected with existing users allowed 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <h3
      class="MuiTypography-root makeStyles-title-3 MuiTypography-h3 MuiTypography-colorTextPrimary"
    >
      Audience
    </h3>
    <table
      class="MuiTable-root"
    >
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Platform
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            calypso
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            User Type
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            All users (new + existing)
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Variations
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
          >
            <table
              class="MuiTable-root"
            >
              <thead
                class="MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    role="columnheader"
                    scope="col"
                  >
                    Name
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    role="columnheader"
                    scope="col"
                  >
                    Percent
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root"
              >
                <tr
                  class="MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    test
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    40
                    %
                  </td>
                </tr>
                <tr
                  class="MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    control
                    <div
                      class="MuiChip-root makeStyles-default-4"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        Default
                      </span>
                    </div>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    60
                    %
                  </td>
                </tr>
              </tbody>
            </table>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Segments
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            No segments assigned
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`renders as expected with no segment assignments 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <h3
      class="MuiTypography-root makeStyles-title-1 MuiTypography-h3 MuiTypography-colorTextPrimary"
    >
      Audience
    </h3>
    <table
      class="MuiTable-root"
    >
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Platform
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            calypso
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            User Type
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            New users only
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Variations
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
          >
            <table
              class="MuiTable-root"
            >
              <thead
                class="MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    role="columnheader"
                    scope="col"
                  >
                    Name
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    role="columnheader"
                    scope="col"
                  >
                    Percent
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root"
              >
                <tr
                  class="MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    test
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    40
                    %
                  </td>
                </tr>
                <tr
                  class="MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    control
                    <div
                      class="MuiChip-root makeStyles-default-2"
                    >
                      <span
                        class="MuiChip-label"
                      >
                        Default
                      </span>
                    </div>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    60
                    %
                  </td>
                </tr>
              </tbody>
            </table>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            role="cell"
            scope="row"
          >
            Segments
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body"
          >
            No segments assigned
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
